src/BinSkim.Rules/ElfRules/BA3031.EnableSafeStackWithClang.cs

```diff
@@ -0,0 +1,90 @@
+using Microsoft.CodeAnalysis.IL.Sdk;
+using Microsoft.CodeAnalysis.Sarif;
+using Microsoft.CodeAnalysis.Sarif.Driver;
+
+namespace Microsoft.CodeAnalysis.IL.Rules
+{
+    [Export(typeof(Skimmer<BinaryAnalyzerContext>)), Export(typeof(ReportingDescriptor))]
+    public class EnableSafeStackWithClang : ElfBinarySkimmer
+    {
+        // Symbol is the same for both C and C++, despite the "cpp" or "cc" in file name.
+        // Clang V7 - V9: "safestack.cc.o"
+        // Clang V10 - V14: "safestack.cpp.o"
+        private static readonly string[] symbolSafeStack = new string[] { "safestack.cpp.o", "safestack.cc.o" };
+
+        /// <summary>
+        /// BA3031
+        /// </summary>
+        public override string Id => RuleIds.EnableSafeStackWithClang;
+
+        /// <summary>
+        /// The SafeStack instrumentation pass protects programs by implementing two separate program stacks, 
+        /// one for return addresses and local variables, and the other for everything else.
+        /// To enable SafeStack, pass '-fsanitize=safe-stack' flag to both compile and link command lines.
+        /// </summary>
+        public override MultiformatMessageString FullDescription => new MultiformatMessageString { Text = RuleResources.BA3031_EnableSafeStackWithClang_Description };
+
+        protected override IEnumerable<string> MessageResourceNames => new string[] {
+                    nameof(RuleResources.BA3031_Pass),
+                    nameof(RuleResources.BA3031_Error),
+                    nameof(RuleResources.NotApplicable_InvalidMetadata)
+                };
+
+        public override AnalysisApplicability CanAnalyzeElf(ElfBinary target, Sarif.PropertiesDictionary policy, out string reasonForNotAnalyzing)
+        {
+            IELF elf = target.ELF;
+
+            if (elf.Type == FileType.Core || elf.Type == FileType.None || elf.Type == FileType.Relocatable)
+            {
+                reasonForNotAnalyzing = MetadataConditions.ElfIsCoreNoneOrObject;
+                return AnalysisApplicability.NotApplicableToSpecifiedTarget;
+            }
+
+            if (!target.Compilers.Any(c => c.Compiler == ElfCompilerType.Clang && c.Version.Major >= 7))
+            {
+                reasonForNotAnalyzing = MetadataConditions.ElfNotBuiltWithClangV7OrLater;
+                return AnalysisApplicability.NotApplicableToSpecifiedTarget;
+            }
+
+            reasonForNotAnalyzing = null;
+            return AnalysisApplicability.ApplicableToSpecifiedTarget;
+        }
+
+        /// <summary>
+        /// Checks if SafeStack is enabled by Symbols.
+        /// </summary>
+        public override void Analyze(BinaryAnalyzerContext context)
+        {
+            IELF elf = context.ElfBinary().ELF;
+
+            IEnumerable<ISymbolEntry> symbols =
+                ElfUtility.GetAllSymbols(elf).Where(sym => sym.Type == SymbolType.File);
+
+            if (symbols.Any(s => symbolSafeStack.Contains(s.Name)))
```

Comment:
Foreach over the symbols, to avoid a 2x traversal in the pathological case. Fire a result and exit if you meet the first condition. Otherwise fall out and fire the pass result.

Reply:
Changed to not use Linq and just use foreach.
(It may not result in 2x traversal though)